[ZEPPELIN-1736] Introduce trash & enable removing folder

tae-jun opened the pull request:
### What is this PR for?
![image](https://cloud.githubusercontent.com/assets/8201019/20925378/03a722f8-bbfa-11e6-9afc-c62cfe6b0e2a.png)
This PR introduces the trash! (which is essential, I think ^^;)

I added basic operations of the trash:
#### Added operations of trash
* `Move` a folder to the trash
* `Move` a note to the trash
* `Restore` a folder
* `Restore` a note
* `Remove` a folder in the trash for good
* `Remove` a note in the trash for good
* `Empty` the trash
* `Restore all` in the trash
* `Move` a note to the trash from the notebook view
* `Remove` a note to the trash for good from the notebook view
* If there is a folder that has the same name, suffix current data rather than just merging(please see screenshot below!)


### What type of PR is it?
[Feature]

### Future work
Maybe it would be better if notebook view notices that a note is in the trash.

### What is the Jira issue?
[ZEPPELIN-1736](https://issues.apache.org/jira/browse/ZEPPELIN-1736)

### How should this be tested?
* Build this PR
* Move some folders or notes to the trash
* Restore them
* Empty the trash
* Restore all notes in the trash
* Remove folders or notes in the trash
* Remove a folder which has the same folder name in the trash

### Screenshots (if appropriate)
## Move, restore, remove a note from the main page
![note_from_main](https://cloud.githubusercontent.com/assets/8201019/20925760/13b0d91c-bbfc-11e6-94c5-2211c9d9a282.gif)

## Move, restore, remove a folder from the main page
![folder_test](https://cloud.githubusercontent.com/assets/8201019/20925774/2884abde-bbfc-11e6-9ba9-7b1f8a935a41.gif)

## Move, remove a note from the notebook view
![noteview_test](https://cloud.githubusercontent.com/assets/8201019/20925837/83564162-bbfc-11e6-9a92-c6bc1984b30d.gif)

## Restore, empty the trash
![restore_empty_trash](https://cloud.githubusercontent.com/assets/8201019/20925841/8f6d78d0-bbfc-11e6-90a9-6f67c95371a5.gif)

## Suffix current date if the folder exists in the trash
![same_folder_name_test](https://cloud.githubusercontent.com/assets/8201019/20926085/e8cb6954-bbfd-11e6-9cdf-acf9b785d833.gif)

### Questions:
* Does the licenses files need update? NO
* Is there breaking changes for older versions? NO
* Does this needs documentation? NO

marchpig commented on zeppelin-web/src/app/home/home.html:
```diff
@@ -32,13 +33,24 @@
     <a style="text-decoration: none;">
-      <i style="font-size: 13px; margin-left: 2px; cursor: pointer; text-decoration: none;"
+      <!-- if note is not in trash -->
+      <i ng-if="!node.isTrash" style="font-size: 13px; margin-left: 2px; cursor: pointer; text-decoration: none;"
```
How about creating a `class` for this `style` into "home.css" file?
Same `style` properties are being used in this file several times.

1ambda commented:
Let me review and give some feedbacks :) Thanks for implementing the useful feature!

marchpig commented on zeppelin-web/src/app/notebook/notebook-actionBar.html:
```diff
@@ -126,11 +126,20 @@
 <!-- put the delete action by itself for your protection -->
       <span class="labelBtn" style="vertical-align:middle; display:inline-block;">
-        <button type="button"
+        <button ng-if="note.name.split('/')[0] === '~Trash'"
```
How about creating a function for this comparison into a "controller.js"?
And it would be nice to make a `const` value for `~Trash`!

tae-jun commented:
Thanks, @1ambda 😄

tae-jun commented:
![image](https://cloud.githubusercontent.com/assets/8201019/20929471/8897948a-bc0d-11e6-941b-4123bc09c8a7.png)
It failed only at Selenium test twice. I guess it's because I **changed the action of remove note button** of notebook view to moving to trash.

But I have no idea where to fix! 😢 Please help!

tae-jun commented:
@marchpig Thanks for your prompt and high-quality review!

You made really valid points. I agree with those at all.

I chose '~' as prefix because of sort order, but I didn't consider Korean. Thanks!

I will fix what you mentioned when I have some time and ping you again 😄 
Thanks again! 👍

cuspymd commented on zeppelin-server/src/main/java/org/apache/zeppelin/socket/NotebookServer.java:
```diff
@@ -894,6 +930,122 @@ private void removeNote(NotebookSocket conn, HashSet<String> userAndRoles,
     broadcastNoteList(subject, userAndRoles);
   }
 
+  private void removeFolder(NotebookSocket conn, HashSet<String> userAndRoles,
```
Is 'Set' interface better than 'HashSet' class as the argument type ?

tae-jun commented:
Thanks for the review @cuspymd! I think so, but I used HashSet because of consistency. Other methods used HashSet.

I think if we want to change to 'Set', wouldn't it be better to open new refactoring PR that change all other 'HashSet' to 'Set' for consistency? 😄 

What do you think?

cuspymd commented:
@tae-jun Yes. Good Idea~!!

tae-jun commented:
Thanks @cuspymd 😄

tae-jun commented:
@marchpig Sorry for my late response! I changed what you addressed except one thing.

### Korean note order
![image](https://cloud.githubusercontent.com/assets/8201019/21017141/b8b797e2-bdab-11e6-94db-7e1bac10389b.png)
### Remove date string when restore
![remove_date](https://cloud.githubusercontent.com/assets/8201019/21017265/16fa00d8-bdac-11e6-876f-c4c403f2192d.gif)

The one thing is about hiding `~Trash` prefix. I definitely agree with you but I think it's not good to make a PR huge. So if you think that feature is essential, I will include it to this PR. But if it's not, I would like to leave that as a future work :)

Actually, I planned to add features related to the trash on notebook view in the future. There are many features need to be implemented!
* Hide `~Trash` prefix
* Notice a note is in the trash
* Disable edit note in the trash (I'm not sure that this is needed, maybe limiting user's behavior)
* Disable renaming note in the trash (not sure as well)

The first and second features are needed in my opinion, so I would like to handle those on next PR!

Please tell me your opinion.
Thanks for the review again. It looks better than before thanks to your comments 😄

marchpig commented:
@tae-jun It works well! Thank you very much for accepting my opinions. :smiley: 
I think that handling `~Trash` prefix in the future is fine to me because it is not a bug but just my opinion :)

However, I found a minor problem.
When a folder name has a trailing space, it is not moved to trash. For example, "My folder /My note".
I think that you can handle this corner case in the future work!

Thanks again for considering my opinions. :+1:

tae-jun commented:
Thanks for your careful review @marchpig 😄 I fixed the bug you found!

It turns out that I didn't normalize folder ID when creating a folder. It could be a serious bug if you didn't find it.

Thanks 👍

tae-jun commented:
### CI is green! 😄 
I **changed Selenium test codes** affected by my PR and it works now.

#### Failed reason
`remove note` button changed to `move to trash` but test code still wanted to click `remove note`. So I made it click `move to trash` button and now it works :)

Please review!
Thanks

Leemoonsoo commented:
@tae-jun thanks for the great contribution!
LGTM and merge to master if there're no further discussions.

tae-jun commented:
Thanks @Leemoonsoo 😄

AhyoungRyu commented:
@tae-jun Nice feature indeed!
Most of cases are working like a charm. 
But as @marchpig said above, 
![screen shot 2016-12-14 at 12 28 46 pm](https://cloud.githubusercontent.com/assets/10060731/21168719/bff892d4-c1f9-11e6-8644-9ced7a81d955.png)
is this intended result? I removed the `note1` and `note2` separately, and the time stamped only for latter one: `NoteDirB/note2`.

And another really minor thing is
![screen shot 2016-12-14 at 12 41 20 pm](https://cloud.githubusercontent.com/assets/10060731/21168885/dd423fb0-c1fa-11e6-8b7f-2a5337497a79.png)

As you can see there are two trash can icons and their sizes are different now. Hmm.. so I would suggest you to use another icon for `Empty trash`(not sure which one will be proper for this..kk) since the trash can icon is already used for `Move note to trash` / `Move folder to trash` / `Remove note permanently` / `Remove folder permanently`.

tae-jun commented:
@AhyoungRyu Always thanks for your nice review 😄 

Yes, it's intended. A time stamp is generated only if a `folder` already exists in the trash. But it's not generated when moving a `note` to the trash.

I made this feature because of a **use case** like this:
### A use case needs time stamp
1. A user moves "NoteDirA" folder to the trash
2. And the user create a folder which has the same name ("NoteDirA")
3. After a month from then, the user forgot the existence of the folder already exists in the trash, and moves the folder to the trash which created later
4. But it was a mistake and the user wants to restore the folder created later

### Necessity of the time stamp for the use case above

**If there is no time stamp:**
The folder created later will be merged into old one and the folder "NoteDirA" will be messy.

#### If there is a time stamp:
The folder created later will be moved to the trash with a time stamp postfix. There will be two folders which are "NoteDirA" (old one) and "NoteDirA removed at 2016-12-14 xx:xx:xx" (new one). In this case, the user can restore the folder successfully.

This is why I added time stamp! :-) Please tell me if there is any other idea.

### Remove permanently icon candidates (?)
I agree with you so I searched Font Awesome and found some candidates for the icon. I think not only `empty trash` but also `Remove note permanently` and `Remove folder permanently` should be replaced together. 

![image](https://cloud.githubusercontent.com/assets/8201019/21180874/543a32b0-c23e-11e6-9f5f-25a0fa25bb84.png)
![image](https://cloud.githubusercontent.com/assets/8201019/21180952/c8ce421a-c23e-11e6-9297-f7df93a36ef5.png)
![image](https://cloud.githubusercontent.com/assets/8201019/21180965/e9a051ea-c23e-11e6-96b8-7df5469a87cf.png)
![image](https://cloud.githubusercontent.com/assets/8201019/21181004/179fe736-c23f-11e6-9e58-36ade60959c4.png)

The icons above are the candidates I found, but please let me know if you have a better one!

Which icon do you prefer? 😄

Thanks for the review again 👍

AhyoungRyu commented:
@tae-jun Appreciate for your precise explanation and picking proper candidates as well :) Tested again to check the time stamp,  it makes sense. Then how about attaching the time stamp for every cases to keep consistency? Currently the firstly removed one doesn't show the stamp. Surely It's just my opinion. 
And regarding the icon, I couldn't find the better icon that has meaning "Clear" either. So you can use anything that you suggested!

tae-jun commented:
I changed `empty trash`, `remove folder permanently`, and `remove note permanently` icons to `X`. I'm not sure this is the best icon for it ^^; but we can change later when we find a suitable icon.
#### Empty trash icon:
![image](https://cloud.githubusercontent.com/assets/8201019/21286529/11d306dc-c49a-11e6-9960-a87f7c4d2f1c.png)
#### Remove folder permanently icon:
![image](https://cloud.githubusercontent.com/assets/8201019/21286535/21cb26a0-c49a-11e6-87d0-1e592484af1f.png)
#### Remove note permanently icon:
![image](https://cloud.githubusercontent.com/assets/8201019/21286537/2b29c4fe-c49a-11e6-96b8-ec9cc771aa55.png)

And I've been thinking about attaching the time stamp to all, but there are some reasons I worry about:
* It makes folder name too long. It's not good to see. So, I also removed infix ` removed at `. 
* MacOS also only attaches the time stamp to duplicated folder names. So, only attaching the time stamp to duplicated ones is more familiar to users, I think.

Now, a duplicated folder name looks like this:
![image](https://cloud.githubusercontent.com/assets/8201019/21286783/b912dca0-c4a0-11e6-8ce9-04b4c9d3c40e.png)

What do you think @AhyoungRyu?
Thanks for the review 😄

1ambda commented:
Hi all, This PR looks like unpopular from the point of view design. I will close this issue.

1ambda commented:
@tae-jun 

It works well as described and here are two things to consider. 

1. Trash is represented in `~Trash` in the Navbar.

<img width="149" alt="screen shot 2016-12-19 at 1 48 36 pm" src="https://cloud.githubusercontent.com/assets/4968473/21301739/91cfc5dc-c5f3-11e6-92c4-4d1010024aae.png">



2. Do we really need *rename*, *clear output* button on notes in Trash? IMO, we need only 2 options. **remove** or **restore** for that notes.

<img width="223" alt="screen shot 2016-12-19 at 1 47 39 pm" src="https://cloud.githubusercontent.com/assets/4968473/21301709/5901b012-c5f3-11e6-9699-f036fc3de694.png">

AhyoungRyu commented:
@tae-jun Thanks for your precise explanation. Yeah it makes sense. And tested again and it looks good for what we talked about :) 

It'll be nice to be merged after addressing @1ambda's feedback as well!

tae-jun commented:
@AhyoungRyu Always thanks for the review :) Also thanks for your understanding.

@1ambda  Thanks for the review! I appreciate it and sorry for my late response.
That was a really nice point. I fixed it as you see below:

<img src="https://cloud.githubusercontent.com/assets/8201019/21429799/b678e794-c8a3-11e6-851f-04d1ad55998e.png" alt="image" width=200>
I tested on Chrome and Firefox. Please check!


As to the second point that you mentioned, I thought a lot of about it as well :)
My opinion is same with you, but I think maybe other users can feel limited.
So I would like to make it configurable as three levels on other PR:
1. super strict: not even open note
2. read only: only can open note and read it
3. unlimited access: able to rename, read, write, and run note

Would you mind if I do that? Please let me know 😄
Thanks for the review again.

1ambda commented:
Hi @tae-jun 

It's good idea to introduce limitation when notes are in trash. Let me comment

(1) what you mentioned is about restricting modification of archived notes (i mean, notes are in trash)
(2) Removing `rename` and `clear output` **does not** mean disabling those functionality. It's just about hiding the menus. We can rename and clear output in the trashed notes. It's not about (1)

So I I think removing `rename` and `clear output` can be handled in this PR.

tae-jun commented:
@1ambda Thanks for your prompt response!
I see what you mean now. Thanks for the explanation :)

Now, only `remove permanently` and `restore` buttons are shown.
<img src="https://cloud.githubusercontent.com/assets/8201019/21433976/7c735306-c8b5-11e6-92a2-8f09ab84e823.png" width=200>
<img src="https://cloud.githubusercontent.com/assets/8201019/21433985/8334b02c-c8b5-11e6-87bf-ada0f978d69b.png" width=200>

How does it look? 😄

1ambda commented:
@tae-jun great! :)

tae-jun commented:
Thanks @1ambda 😄

AhyoungRyu commented:
Finally CI is green now! LGTM 🎅🏼

tae-jun commented:
@AhyoungRyu Yeah! Thanks :) Merry Christmas~

tae-jun commented:
Conflict resolved 😄

Leemoonsoo commented:
LGTM!

AhyoungRyu commented:
@tae-jun Thanks for your great work! Merge if there are no more discussions :)

soralee commented:
@tae-jun It's great feature! So, I would like to suggest one thing.
Currently, the filter function in Zeppelin homepage doesn't work as expected and this issue is handled in [ZEPPELIN-1864](https://issues.apache.org/jira/browse/ZEPPELIN-1864).

But, when using filter function, all icon is disappeared like below.
![z1736](https://cloud.githubusercontent.com/assets/8110458/21514983/d335bf50-cd0e-11e6-8297-a48bbacdf308.gif)
At this time, I think It would be nice to show some icon next to note or folder if you don't mind.
What do you think?

tae-jun commented:
Thanks for the review, @soralee! That was a big deal.

I pushed a commit and now it looks like this:
<img src="https://cloud.githubusercontent.com/assets/8201019/21564175/bc5fc4c2-cecd-11e6-88dd-d981659fb63e.png" width="400">
I just made it show trash icon if it's in the trash.

And it looks like that the search result only shows notes, not folders.
It would be great if it's handled on the issue that you said :)

After resolving the issue, the trash folder should show trash icon.
I can do that if you want! Please assign that issue to me if you want :)

What do you think? 😄

soralee commented:
Thanks for considering my suggestion, @tae-jun 😄 
It is good idea and I agree with you! 
I think this problem is a little too hard to solve.
So, anyway, I will test again!

AhyoungRyu commented:
I think it's good to merge. So let's improve after then! 
Will merge this one if there are no further discussions. @tae-jun Appreciate for this AWESOME feature 👍

tae-jun commented:
Thanks @AhyoungRyu 😄

AhyoungRyu commented:
@tae-jun I tried to run `npm run test` under `zeppelin-web` in my local, but it failed with below error. 
```
PhantomJS 2.1.1 (Mac OS X 0.0.0) Factory: NoteList should generate both flat list and folder-based list properly FAILED
	TypeError: undefined is not an object (evaluating 'note.name.split') in src/index.js (line 8407)
	src/index.js:8407:36
	arrayMap@bower_components/lodash/lodash.js:1431:33
	map@bower_components/lodash/lodash.js:6693:18
	setNotes@src/index.js:8406:32
	test/spec/factory/noteList.js:78:23
PhantomJS 2.1.1 (Mac OS X 0.0.0): Executed 69 of 71 (1 FAILED) (0 secs / 0.4LOG: function (columns, rows, comment) { ... }
PhantomJS 2.1.1 (Mac OS X 0.0.0): Executed 69 of 71 (1 FAILED) (0 secs / 0.4PhantomJS 2.1.1 (Mac OS X 0.0.0): Executed 70 of 71 (1 FAILED) (0 secs / 0.4LOG: function (columns, rows, comment) { ... }
PhantomJS 2.1.1 (Mac OS X 0.0.0): Executed 70 of 71 (1 FAILED) (0 secs / 0.4PhantomJS 2.1.1 (Mac OS X 0.0.0): Executed 71 of 71 (1 FAILED) (0 secs / 0.4PhantomJS 2.1.1 (Mac OS X 0.0.0): Executed 71 of 71 (1 FAILED) (0.334 secs / 0.424 secs)
```

Can you take a look again?

tae-jun commented on zeppelin-web/src/components/noteListDataFactory/noteList.datafactory.js:
```diff
@@ -15,15 +15,20 @@
       setNotes: function(notesList) {
         // a flat list to boost searching
-        notes.flatList = angular.copy(notesList);
+        notes.flatList = _.map(angular.copy(notesList), (note) => {
```
@felizbear Thanks! You are right! 👍  I will push a commit soon

tae-jun commented:
@AhyoungRyu Thanks!!! I fixed the test error thanks to you 👍 

I guess CI will be green this time 😄

AhyoungRyu commented:
1 build failed with below error, but it's irrelevant to this change. 

```
[ERROR] Failed to execute goal org.apache.maven.plugins:maven-surefire-plugin:2.17:test (default-test) on project zeppelin-pig: Execution default-test of goal org.apache.maven.plugins:maven-surefire-plugin:2.17:test failed: The forked VM terminated without properly saying goodbye. VM crash or System.exit called?
[ERROR] Command was /bin/sh -c cd /home/travis/build/apache/zeppelin/pig && /usr/lib/jvm/java-7-oracle/jre/bin/java -Xmx2g -Xms1g -Dfile.encoding=UTF-8 -jar /home/travis/build/apache/zeppelin/pig/target/surefire/surefirebooter300784690996837886.jar /home/travis/build/apache/zeppelin/pig/target/surefire/surefire6525093842604398164tmp /home/travis/build/apache/zeppelin/pig/target/surefire/surefire_327671988278855013716tmp
[ERROR] -> [Help 1]
[ERROR] 
[ERROR] To see the full stack trace of the errors, re-run Maven with the -e switch.
[ERROR] Re-run Maven using the -X switch to enable full debug logging.
[ERROR] 
[ERROR] For more information about the errors and possible solutions, please read the following articles:
[ERROR] [Help 1] http://cwiki.apache.org/confluence/display/MAVEN/PluginExecutionException
[ERROR] 
[ERROR] After correcting the problems, you can resume the build with the command
[ERROR]   mvn <goals> -rf :zeppelin-pig
``` 

So let me merge into master in this time. Thanks for your long long patience and effort @tae-jun 🌄
